src/runtime_bignum_test.nr

```diff
@@ -3,6 +3,194 @@ use crate::runtime_bignum::BigNumInstance;
 use crate::runtime_bignum::BigNumParamsTrait;
 use crate::utils::u60_representation::U60Repr;
 use crate::fields::bn254Fq::BNParams as BNParams;
+use crate::fields::secp256k1Fq::Secp256k1_Fq_Params;
+use crate::fields::bls12_381Fq::BLS12_381_Fq_Params;
+
+//             
+
+#[test]
+fn test_bls() {
+    let X1: BigNum<4, BLS12_381_Fq_Params> = BigNum {
+        limbs: [
+            0x55e83ff97a1aeffb3af00adb22c6bb, 0x8c4f9774b905a14e3a3f171bac586c, 0xa73197d7942695638c4fa9ac0fc368, 0x17f1d3
+        ]
+    };
+    X1.validate_in_field();
+    let mut (_, XX_mul_3): (BigNum<4, BLS12_381_Fq_Params>, BigNum<4, BLS12_381_Fq_Params> ) = unsafe {
+        BigNum::__compute_quadratic_expression(
+            [[X1, X1, X1]],
+            [[false, false, false]],
+            [[X1]],
+            [[false]],
+            [],
+            []
+        )
+    };
+    XX_mul_3.validate_in_field();
+}
+comptime fn make_test(f: StructDefinition, N: Quoted, typ: Quoted) -> Quoted {
+    let k = f.name();
+    quote{
+impl $k {
+#[test]
+fn test_add() {
+    let bn = $typ ::get_instance();
+    let a: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([1, 2, 3, 4]) };
+    let b: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([4, 5, 6, 7]) };
+    let one: BigNum<$N, $typ> = BigNum::one();
+    a.validate_in_range();
+    bn.validate_in_field(a);
+    b.validate_in_range();
+    bn.validate_in_field(b);
+
+    let mut c = bn.add(a, b);
+    c = bn.add(c, c);
+    let d = bn.mul(bn.add(a, b), bn.add(one, one));
+    assert(bn.eq(c, d));
+
+    let e = bn.add(one, one);
+    for i in 1..$N {
+        assert(e.limbs[i] == 0);
+    }
+    assert(e.limbs[0] == 2);
+}
+    
+
+#[test]
+fn test_mul() {
+    let bn = $typ ::get_instance();
+    let a: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([1, 2, 3, 4]) };
+    let b: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([4, 5, 6, 7]) };
+
+    let c = bn.mul(bn.add(a, b), bn.add(a, b));
+    let d = bn.add(
+        bn.add(
+            bn.add(bn.mul(a, a), bn.mul(b, b)),
+            bn.mul(a, b)
+        ),
+        bn.mul(a, b)
+    );
+    assert(bn.eq(c, d));
+}
+
+#[test]
+fn test_quadratic_expression() {
+    let bn = $typ ::get_instance();
+    for i in 0..32 {
+    let X1: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([i as u8,2,3,4]) };
+    let Y1: BigNum<$N, $typ> = unsafe{ bn.__derive_from_seed([i as u8,i as u8,6,7]) };
+    let Z1: BigNum<$N, $typ> = BigNum::one();
+
+        let (_, YY_mul_2): (BigNum<$N, $typ>, BigNum<$N, $typ> ) = unsafe{BigNum::__compute_quadratic_expression([[Y1]], [[false]], [[Y1, Y1]], [[false, false]], [], [])};
```

Comment:
file could use a `nargo fmt`. Although, I know the formatter struggles with macros.